File: csharp/robot-simulator/RobotSimulator.cs
enum Bearing
{
  North, East, South, West
}

class Coordinate
{
  public Coordinate(int x, int y)
  {
    X = x;
    Y = y;
  }

  public override bool Equals(object obj)
  {
    var compare = (Coordinate)obj;
    return X == compare.X && Y == compare.Y;
  }

  public void Adjust(Bearing bearing)
  {
    switch (bearing)
    {
      case Bearing.North:
        Y = Y + 1;
        break;
      case Bearing.East:
        X = X + 1;
        break;
      case Bearing.South:
        Y = Y - 1;
        break;
      case Bearing.West:
        X = X - 1;
        break;
    }
  }

  public int X { get; set; }
  public int Y { get; set; }
}

class RobotSimulator
{
  public RobotSimulator(Bearing bearing, Coordinate coordinate)
  {
    Bearing = bearing;
    Coordinate = coordinate;
  }

  public void TurnLeft()
  {
    switch (Bearing)
    {
      case Bearing.North:
        Bearing = Bearing.West;
        break;
      case Bearing.East:
        Bearing = Bearing.North;
        break;
      case Bearing.South:
        Bearing = Bearing.East;
        break;
      case Bearing.West:
        Bearing = Bearing.South;
        break;
    }
  }

  public void TurnRight()
  {
    switch (Bearing)
    {
      case Bearing.North:
        Bearing = Bearing.East;
        break;
      case Bearing.East:
        Bearing = Bearing.South;
        break;
      case Bearing.South:
        Bearing = Bearing.West;
        break;
      case Bearing.West:
        Bearing = Bearing.North;
        break;
    }
  }

  public void Simulate(string commands)
  {
    foreach (var c in commands)
    {
      switch(c)
      {
        case 'R':
          TurnRight();
          break;
        case 'L':
          TurnLeft();
          break;
        case 'A':
          Coordinate.Adjust(Bearing);
          break;
      }
    }
  }

  public Bearing Bearing { get; set; }
  public Coordinate Coordinate { get; set; }
}

File: javascript/robot-simulator/robot-simulator.js
var Robot = function () {
  this.bearing = "north";
  this.coordinates = [0, 0];
};

Robot.prototype.orient = function (direction) {
  if (direction != 'north' && direction != 'east' && direction != 'south' && direction != 'west') {
    throw 'Invalid Robot Bearing';
  }
  this.bearing = direction;
};

Robot.prototype.turnRight = function () {
  switch (this.bearing) {
    case "north":
      this.bearing = "east";
      break;
    case "east":
      this.bearing = "south";
      break;
    case "south":
      this.bearing = "west";
      break;
    case "west":
      this.bearing = "north";
      break;
  }
};

Robot.prototype.turnLeft = function () {
  switch (this.bearing) {
    case "north":
      this.bearing = "west";
      break;
    case "east":
      this.bearing = "north";
      break;
    case "south":
      this.bearing = "east";
      break;
    case "west":
      this.bearing = "south";
      break;
  }
};

Robot.prototype.at = function (x, y) {
  this.coordinates = [x, y];
};

Robot.prototype.advance = function () {
  switch (this.bearing) {
    case "north":
      this.coordinates = [this.coordinates[0], this.coordinates[1] + 1];
      break;
    case "east":
      this.coordinates = [this.coordinates[0] + 1, this.coordinates[1]];
      break;
    case "south":
      this.coordinates = [this.coordinates[0], this.coordinates[1] - 1];
      break;
    case "west":
      this.coordinates = [this.coordinates[0] - 1, this.coordinates[1]];
      break;
  }
};

Robot.prototype.instructions = function (instructions) {
  var converted = [];
  for (var i = 0; i < instructions.length; i++) {
    switch (instructions[i]) {
      case 'A':
        converted.push('advance');
        break;
      case 'R':
        converted.push('turnRight');
        break;
      case 'L':
        converted.push('turnLeft');
        break;
    }
  }
  return converted;
};

Robot.prototype.place = function (location) {
  this.at(location.x, location.y);
  this.orient(location.direction);
};

Robot.prototype.evaluate = function (instructions) {
  var convertedInstructions = this.instructions(instructions);
  for (var i = 0; i < convertedInstructions.length; i++) {
    this[convertedInstructions[i]]();
  }
};

module.exports = Robot;